src/sync/partial_sync.cpp

```diff
@@ -56,6 +83,153 @@ void update_schema(Group& group, Property matches_property)
+    if (existing_matches_property != matches_property)
+        throw std::runtime_error("An existing subscription exists with the same name, but a different result type.");
+
+    return true;
+}
+
+class WorkQueue {
+public:
+    ~WorkQueue()
+    {
+        reset();
+    }
+
+    void reset()
+    {
+        {
+            std::unique_lock<std::mutex> lock(m_mutex);
+            m_stopping = true;
+        }
+        m_cv.notify_one();
+
+        if (m_thread.joinable())
+            m_thread.join();
+
+        {
+            std::unique_lock<std::mutex> lock(m_mutex);
+            m_stopping = false;
+        }
+    }
+
+    void enqueue(std::function<void()> function)
+    {
+        {
+            std::unique_lock<std::mutex> lock(m_mutex);
+            m_queue.push_back(std::move(function));
+
+            if (m_stopped)
+                create_thread();
+        }
+        m_cv.notify_one();
+    }
+
+private:
+    void create_thread()
+    {
+        if (m_thread.joinable())
+            m_thread.join();
+
+        m_thread = std::thread([this] {
+            std::vector<std::function<void()>> queue;
+
+            std::unique_lock<std::mutex> lock(m_mutex);
+            while (!m_stopping &&
+                   m_cv.wait_for(lock, std::chrono::milliseconds(500),
+                                 [&] { return !m_queue.empty() || m_stopping; })) {
+
+                swap(queue, m_queue);
+
+                lock.unlock();
+                for (auto& f : queue)
+                    f();
+                queue.clear();
+                lock.lock();
+            }
+
+            m_stopped = true;
+        });
+        m_stopped = false;
+    }
+
+    std::mutex m_mutex;
+    std::condition_variable m_cv;
+    std::vector<std::function<void()>> m_queue;
+    std::thread m_thread;
+    bool m_stopping = false;
+    bool m_stopped = true;
+};
+
+static auto& work_queue = *new WorkQueue;
+
+void async_register_query(Realm& realm, std::string object_type, std::string query, std::string name,
+                          std::function<void(std::exception_ptr)> callback)
+{
+    work_queue.enqueue([object_type=std::move(object_type), query=std::move(query), name=std::move(name),
+                        callback=std::move(callback), config=realm.config()] {
+        auto realm = Realm::get_shared_realm(config);
+        realm->begin_transaction();
+        auto cleanup = util::make_scope_exit([&]() noexcept {
+            if (realm->is_in_transaction())
+                realm->cancel_transaction();
+        });
+
+        auto matches_property = std::string(object_type) + "_matches";
+
+        update_schema(realm->read_group(), Property(matches_property, PropertyType::Object|PropertyType::Array, object_type));
+        auto result_sets_schema = std::make_unique<ObjectSchema>(realm->read_group(), result_sets_type_name);
```

Comment:
This doesn't appear to need to be heap allocated.

Reply:
Quite right. An earlier version heap allocated it to ensure it lived as long as an `Object` that used it, but that's no longer a concern.